e2e/LongHaul/service/ModuleOperations.cs

```diff
@@ -0,0 +1,132 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using Mash.Logging;
+using Newtonsoft.Json;
+using static Microsoft.Azure.Devices.LongHaul.Service.LoggingConstants;
+
+namespace Microsoft.Azure.Devices.LongHaul.Service
+{
+    internal class ModuleOperations
+    {
+        private readonly IotHubServiceClient _serviceClient;
+        private readonly string _deviceId;
+        private readonly string _moduleId;
+        private readonly Logger _logger;
+
+        private long _totalMethodCallsCount = 0;
+        private long _totalDesiredPropertiesUpdatesCount = 0;
+
+        private static readonly TimeSpan s_directMethodInvokeInterval = TimeSpan.FromSeconds(3);
+        private static readonly TimeSpan s_desiredPropertiesSetInterval = TimeSpan.FromSeconds(3);
+
+        private static readonly TimeSpan s_retryInterval = TimeSpan.FromSeconds(1);
+
+        public ModuleOperations(IotHubServiceClient serviceClient, string deviceId, string moduleId, Logger logger)
+        {
+            _serviceClient = serviceClient;
+            _deviceId = deviceId;
+            _moduleId = moduleId;
+            _logger = logger;
+            _logger.LoggerContext.Add("DeviceId", deviceId);
+        }
+
+        public async Task InvokeDirectMethodAsync(Logger logger, CancellationToken ct)
+        {
+            logger.LoggerContext.Add(OperationName, DirectMethod);
+            Stopwatch sw = new();
+            while (!ct.IsCancellationRequested)
+            {
+                var payload = new CustomDirectMethodPayload
+                {
+                    RandomId = Guid.NewGuid(),
+                    CurrentTimeUtc = DateTimeOffset.UtcNow,
+                    MethodCallsCount = ++_totalMethodCallsCount,
+                };
+
+                var methodInvocation = new DirectMethodServiceRequest("EchoPayload")
+                {
+                    Payload = payload,
+                    ResponseTimeout = TimeSpan.FromSeconds(30),
+                };
+
+                logger.Trace($"Invoking direct method for device: {_deviceId}, module: {_moduleId}", TraceSeverity.Information);
+                logger.Metric(TotalDirectMethodCallsCount, _totalMethodCallsCount);
```

Comment:
Similarly, modify metric names in the context of "device" and "module" accordingly.